modules/core/image/src/Plugin/ImageEffect/AutoOrientImageEffect.php

```diff
@@ -0,0 +1,217 @@
+        break;
+
+      case 4:
+        $result = $image->apply('mirror', ['y_axis' => TRUE]);
+        break;
+
+      case 5:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 270]);
+        }
+        break;
+
+      case 6:
+        $result = $image->apply('rotate', ['degrees' => 90]);
+        break;
+
+      case 7:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 90]);
+        }
+        break;
+
+      case 8:
+        $result = $image->apply('rotate', ['degrees' => 270]);
+        break;
+
+      default:
+        $result = TRUE;
+    }
+
+    // If the transformation failed, bail.
+    if (!$result) {
+      $this->logger->error('Image auto-orient failed using the %toolkit toolkit on %path (%mimetype)', [
+        '%toolkit' => $image->getToolkitId(),
+        '%path' => $image->getSource(),
+        '%mimetype' => $image->getMimeType(),
+      ]);
+      return $result;
+    }
+
+    // Return the result.
+    return $result;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transformDimensions(array &$dimensions, $uri) {
+
+    // Only proceed if the exif PHP extension is loaded.
+    if (!extension_loaded('exif')) {
+      return;
+    }
+
+    // This logic is adapted from the Drupal Image Effects module.
+    // @see https://git.drupalcode.org/project/image_effects/-/blob/730a1e6f5f947ef4d6eec5d359c6968ce93594f9/src/Plugin/ImageEffect/AutoOrientImageEffect.php
+    $dimensions['width'] = $dimensions['width'] ? (int) $dimensions['width'] : NULL;
+    $dimensions['height'] = $dimensions['height'] ? (int) $dimensions['height'] : NULL;
+    if ($dimensions['width'] && $dimensions['height']) {
+      $orientation = $this->loadExifOrientation($uri);
+      if (in_array($orientation, [5, 6, 7, 8])) {
+        $tmp = $dimensions['width'];
+        $dimensions['width'] = $dimensions['height'];
+        $dimensions['height'] = $tmp;
+      }
+    }
+  }
+
+  /**
+   * Load EXIF Orientation value from an image file.
+   *
+   * @param string $uri
+   *   The image file URI.
+   *
+   * @return int|null
+   *   Returns an integer representing the EXIF Orientation value, or NULL if no
+   *   Orientation key was found.
+   */
+  protected function loadExifOrientation($uri) {
+
+    // Get the image file path.
+    $path = $this->fileSystem->realpath($uri);
+
+    // Attempt to load EXIF data.
+    $exif = @exif_read_data($path);
+
+    // Return the Orientation value, if available.
+    if (is_array($exif) && isset($exif['Orientation'])) {
+      return $exif['Orientation'];
+    }
+
+    // Otherwise return NULL.
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    return $form;
+  }
```

Comment:
Since this effect doesn't use any configuration, could we extend directly from `ImageEffectBase`?

If we did have configuration I think we would need to define the `image.effect.farm_image_auto_orient` schema, but without configuration I think we're OK. I see the core `image_module_test` only provides a schema for the configurable image effect it provides, but the non configurable effects are not included: https://git.drupalcode.org/project/drupal/-/blob/10.5.x/core/modules/image/tests/modules/image_module_test/config/schema/image_module_test.schema.yml?ref_type=heads

Reply:
Ah good point! Fixup: https://github.com/farmOS/farmOS/pull/941/commits/c105ca32137c7811653842d808c69b34821c4900